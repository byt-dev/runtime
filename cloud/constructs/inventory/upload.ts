import * as cdk from 'aws-cdk-lib'
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { Construct } from 'constructs'
import { TypeScriptCode } from '@mrgrain/cdk-esbuild'
import * as path from 'path'
import { LambdaConfig } from '../lambda-config'
export interface UploadApiProps {
  readonly bucket: cdk.aws_s3.IBucket
  readonly cognito: {
    readonly userPoolId: string
    readonly clientId: string
    readonly domain: string
    readonly region: string
  }
}

export class UploadApi extends Construct {
  public readonly distribution: cdk.aws_cloudfront.Distribution

  constructor(scope: Construct, id: string, props: UploadApiProps) {
    super(scope, id)

    const { bucket } = props
    const code = new TypeScriptCode(path.join(__dirname, 'edge/upload.origin-request.ts'), {
      buildOptions: {
        format: "cjs",
        outfile: "index.js",
      },
    })

    const fn = new cdk.aws_cloudfront.experimental.EdgeFunction(this, "handler", {
      code,
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      timeout: cdk.Duration.seconds(10),
      memorySize: 1024,
    });


    const codeResponse = new TypeScriptCode(path.join(__dirname, 'edge/upload.origin-response.ts'), {
      buildOptions: {
        format: "cjs",
        outfile: "index.js",
    },
    })

    const fnResponse = new cdk.aws_cloudfront.experimental.EdgeFunction(this, "response.handler", {
      code: codeResponse,
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      timeout: cdk.Duration.seconds(10),
      memorySize: 1024,
    });

    const lambda = new LambdaConfig(this, 'LambdaConfig', {
      function: fn,
      config: {
        cognito: props.cognito
      },
    })

    const myHostedZone = cdk.aws_route53.HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
      hostedZoneId: 'Z063149824WBMACIEVUV5',
      zoneName: 'run.byt.dev',
    });

    // Create "upload.byt.dev" cname
    const domain = new cdk.aws_route53.CnameRecord(this, 'domain', {
      zone: myHostedZone,
      recordName: 'upload',
      domainName: 'run.byt.dev',
    })

    const cert = new cdk.aws_certificatemanager.Certificate(this, 'Certificate', {
      domainName: domain.domainName,
      certificateName: 'Byt Upload Service', // Optionally provide an certificate name
      validation: cdk.aws_certificatemanager.CertificateValidation.fromDns(myHostedZone),
    });

    const identity = new cdk.aws_cloudfront.OriginAccessIdentity(this, 'OAI')
    bucket.grantReadWrite(identity)

    // cloudfront distribution
    this.distribution = new cdk.aws_cloudfront.Distribution(this, 'Distribution', {
      defaultBehavior: {
        origin: new cdk.aws_cloudfront_origins.S3Origin(bucket, {originAccessIdentity: identity}),
        allowedMethods: cdk.aws_cloudfront.AllowedMethods.ALLOW_ALL,
        originRequestPolicy: cdk.aws_cloudfront.OriginRequestPolicy.ALL_VIEWER_AND_CLOUDFRONT_2022,
        edgeLambdas: [
          {
            includeBody: false,
            functionVersion: lambda.version,
            eventType: cdk.aws_cloudfront.LambdaEdgeEventType.ORIGIN_REQUEST,
          },
          {
            includeBody: false,
            functionVersion: fnResponse.currentVersion,
            eventType: cdk.aws_cloudfront.LambdaEdgeEventType.ORIGIN_RESPONSE,
          }
        ]
      },
      certificate: cert,
      domainNames: [domain.domainName],
      priceClass: cdk.aws_cloudfront.PriceClass.PRICE_CLASS_100,
    });

  }
}