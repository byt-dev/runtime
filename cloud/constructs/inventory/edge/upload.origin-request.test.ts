import payloadType from './upload.origin-request.payload.json'
import { handler } from './upload.origin-request'
import * as fs from "fs";
import * as path from "path";

const ulidValue = "01F1ZQZJ1XQZJ0XQZJ1XQZJ1XQ"

jest.mock("ulid", () => ({
  ulid: jest.fn().mockReturnValue('01F1ZQZJ1XQZJ0XQZJ1XQZJ1XQ'),
}));

jest.mock("jose", () => ({
  createLocalJWKSet: jest.fn().mockReturnValue({}),
  jwtVerify: jest.fn().mockReturnValue({ payload: { sub: 'user', username: 'test-foobar' } }),
}));

// @ts-ignore
global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({}),
  })
);

describe('upload.origin-request', () => {
  let payload: typeof payloadType;

  beforeEach(() => {
    payload = JSON.parse(fs.readFileSync(path.join(__dirname, "upload.origin-request.payload.json"), "utf-8")) as any
  })

  it('works for PUT request', async () => {
    const result = await handler(payload.event, payload.context)
    expect(result).toMatchSnapshot()
  })

  it('denies GET request', async () => {
    const { event, context } = {...payload}
    event.Records[0].cf.request.method = 'GET'
    const result = await handler(event, context)
    expect(result).toMatchSnapshot()
  })

  it('denies POST request', async () => {
    const { event, context } = {...payload}
    event.Records[0].cf.request.method = 'POST'
    const result = await handler(event, context)
    expect(result).toMatchSnapshot()
  })

  it ('denies request without "authorization" header', async () => {
    const { event, context } = {...payload}
    delete event.Records[0].cf.request.headers.authorization
    const result = await handler(event, context)
    expect(result).toMatchSnapshot()
  })

  it ('sets "x-amz-acl" header so the file can be accessed properly', async () => {
    const { event, context } = {...payload}
    event.Records[0].cf.request.method = 'PUT'
    const result = await handler(event, context)
    expect(result.headers['x-amz-acl']).toEqual([{ key: 'x-amz-acl', value: 'bucket-owner-full-control' }])
  })

  it ('sets "host" header', async () => {
    const { event, context } = {...payload}
    event.Records[0].cf.request.method = 'PUT'
    const result = await handler(event, context)
    expect(result.headers['host']).toEqual([{ key: 'Host', value: payload.event.Records[0].cf.request.origin.s3.domainName }])
  })

  it ('rewrites the URI to a random ULID', async () => {
    const { event, context } = {...payload}
    const result = await handler(event, context)
    expect(result.uri).toEqual(`/test-foobar/${ulidValue}.gz`)
  })
})