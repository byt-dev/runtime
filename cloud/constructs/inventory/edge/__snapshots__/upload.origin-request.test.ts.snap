// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload.origin-request denies GET request 1`] = `
{
  "body": "Forbidden",
  "headers": {
    "content-type": [
      {
        "key": "Content-Type",
        "value": "text/plain",
      },
    ],
  },
  "status": "403",
  "statusDescription": "Forbidden",
}
`;

exports[`upload.origin-request denies POST request 1`] = `
{
  "body": "Forbidden",
  "headers": {
    "content-type": [
      {
        "key": "Content-Type",
        "value": "text/plain",
      },
    ],
  },
  "status": "403",
  "statusDescription": "Forbidden",
}
`;

exports[`upload.origin-request denies request without "authorization" header 1`] = `
{
  "body": "Forbidden",
  "headers": {
    "content-type": [
      {
        "key": "Content-Type",
        "value": "text/plain",
      },
    ],
  },
  "status": "403",
  "statusDescription": "Forbidden",
}
`;

exports[`upload.origin-request works for PUT request 1`] = `
{
  "clientIp": "79.140.182.36",
  "headers": {
    "accept-encoding": [
      {
        "key": "accept-encoding",
        "value": "br, gzip, deflate",
      },
    ],
    "content-length": [
      {
        "key": "Content-Length",
        "value": "62",
      },
    ],
    "host": [
      {
        "key": "Host",
        "value": "cloudspec-upload-test-sebastian-mybucketf68f3ff0-bnbp1iesevds.s3.us-east-1.amazonaws.com",
      },
    ],
    "user-agent": [
      {
        "key": "User-Agent",
        "value": "undici",
      },
    ],
    "via": [
      {
        "key": "Via",
        "value": "1.1 3e3281afcc193fd3ff8fb81e9d5a5830.cloudfront.net (CloudFront)",
      },
    ],
    "x-amz-acl": [
      {
        "key": "x-amz-acl",
        "value": "bucket-owner-full-control",
      },
    ],
    "x-forwarded-for": [
      {
        "key": "X-Forwarded-For",
        "value": "79.140.182.36",
      },
    ],
  },
  "method": "PUT",
  "origin": {
    "s3": {
      "authMethod": "origin-access-identity",
      "customHeaders": {},
      "domainName": "cloudspec-upload-test-sebastian-mybucketf68f3ff0-bnbp1iesevds.s3.us-east-1.amazonaws.com",
      "path": "",
      "region": "us-east-1",
    },
  },
  "querystring": "",
  "uri": "/test-foobar/01F1ZQZJ1XQZJ0XQZJ1XQZJ1XQ.gz",
}
`;
